daal4py/sklearn/neighbors/knn.py

```diff
@@ -0,0 +1,525 @@
+
+        n_samples_fit = self.n_samples_fit_
+        if n_neighbors > n_samples_fit:
+            raise ValueError(
+                "Expected n_neighbors <= n_samples, "
+                " but n_samples = %d, n_neighbors = %d" %
+                (n_samples_fit, n_neighbors)
+            )
+
+        n_jobs = effective_n_jobs(self.n_jobs)
+        chunked_results = None
+
+        try:
+            fptype = getFPType(X)
+        except ValueError:
+            fptype = None
+
+        if self._fit_method in ['brute', 'kd_tree'] \
+        and (self.effective_metric_ == 'minkowski' and self.p == 2 or self.effective_metric_ == 'euclidean') \
+        and fptype is not None:
+
+            if self._fit_method == 'brute':
+                knn_classification_training = d4p.bf_knn_classification_training
+                knn_classification_prediction = d4p.bf_knn_classification_prediction
+            else:
+                knn_classification_training = d4p.kdtree_knn_classification_training
+                knn_classification_prediction = d4p.kdtree_knn_classification_prediction
+
+            alg_params = {
+                'fptype': fptype,
+                'method': 'defaultDense',
+                'k': n_neighbors,
+                'resultsToCompute': 'computeIndicesOfNeightbors|computeDistances',
+                'resultsToEvaluate': 'none'
+            }
+
+            training_alg = knn_classification_training(**alg_params)
+
+            fit_X = d4p.get_data(self._fit_X)
+            training_result = training_alg.compute(fit_X)
+
+            prediction_alg = knn_classification_prediction(**alg_params)
+
+            X = d4p.get_data(X)
+            prediction_result = prediction_alg.compute(X, training_result.model)
+
+            if return_distance:
+                results = prediction_result.distances, prediction_result.indices
+            else:
+                results = prediction_result.indices
+
+        elif (self._fit_method == 'brute' and
+                self.effective_metric_ == 'precomputed' and issparse(X)):
+            results = _kneighbors_from_graph(
+                X, n_neighbors=n_neighbors,
+                return_distance=return_distance)
+
+        elif self._fit_method == 'brute':
+            reduce_func = partial(self._kneighbors_reduce_func,
+                                  n_neighbors=n_neighbors,
+                                  return_distance=return_distance)
+
+            # for efficiency, use squared euclidean distances
+            if self.effective_metric_ == 'euclidean':
+                kwds = {'squared': True}
+            else:
+                kwds = self.effective_metric_params_
+
+            chunked_results = list(pairwise_distances_chunked(
+                X, self._fit_X, reduce_func=reduce_func,
+                metric=self.effective_metric_, n_jobs=n_jobs,
+                **kwds))
+
+        elif self._fit_method in ['ball_tree', 'kd_tree']:
+            if issparse(X):
```

Comment:
Maybe callback to KNN from sklearn without copy paste?

Reply:
Fallback/call to stock knn is added